# Results 

Now we will look at the conclusions from the analysis and identify the model with the best overall performance. Additionally, we will discuss applications and next steps to use the model. 

In [11]:
!pip3 install pandas
!pip3 install plotly

import plotly.express as px
import pandas as pd


[notice] A new release of pip is available: 23.1.2 -> 23.3.2
[notice] To update, run: pip install --upgrade pip

[notice] A new release of pip is available: 23.1.2 -> 23.3.2
[notice] To update, run: pip install --upgrade pip


In [12]:
%store -r model_comparisons

In [13]:
model_comparisons.head(20)

Model  Train Time  Train Accuracy  Test Accuracy
0           Logistic Regression    0.079745        0.983619       0.947273
1                   KNN (n = 5)    0.017506        0.983619       0.947273
2     Decision Tree (depth = 3)    0.018919        0.781591       0.747273
3                  SVM (linear)    0.095366        1.000000       1.000000
4                   KNN (n = 3)    0.017691        1.000000       1.000000
5                   KNN (n = 5)    0.015609        0.998440       0.994545
6                  KNN (n = 10)    0.014902        0.969579       0.927273
7                  KNN (n = 20)    0.015343        0.836193       0.816364
8                  KNN (n = 50)    0.014757        0.770671       0.747273
9     Decision Tree (depth = 5)    0.020814        0.821373       0.765455
10   Decision Tree (depth = 10)    0.018789        0.886115       0.800000
11   Decision Tree (depth = 50)    0.021296        1.000000       0.938182
12  Decision Tree (depth = 100)    0.021623        1.000000       0.938182

In [14]:
model_traintime_testaccuracy_graph = px.scatter(model_comparisons, x='Train Time', y='Test Accuracy', text='Model',
                 title='Model Performance: Train Time vs Test Accuracy')
model_traintime_testaccuracy_graph.show()

## Best Model 
As seen from the charts above, the top performing models are: 
* KNN (n=5): Train Time = 0.015609, Test Accuracy = 0.994545
* KNN (n=10): Train Time = 0.014902, Test Accuracy = 0.927273
* SVM (linear): Train Time = 0.095366 , Test Accuracy = 1.0
* Decision Tree (depth = 50): Train Time = 0.021296, Test Accuracy = 0.938182

The best tradeoff between runtime and accuracy seems to be one of the KNN models and, considering there is a 0.001s time difference for an accuracy increase from .92 to .99, the KNN model with N=5 is the best overall model. 


## Applications
There are many applications of this model in a software application either for waste management departments in NYC or an application for the resident of NYC. 

In waste sorting and management facilities, this model has the following potential imapacts: 
* The use of this ML model will help to identify categories for complex waste compositions more accurately which will reduce errors in sorting at recycling facilities.
* Reduced manual intervention, optimizing resource allocation within facilities 
* Reduced operational costs by streamlining the sorting process, potentially lowering labor costs and resource wastage
* Quicker and more accurate sorting can expedite recycling processes, increasing throughput and reducing bottlenecks.
* Accurate sorting can also directly improve the quality of recyclable materials, making them more desirable and valuable for resale or reuse in manufacturing. This, inturn, contributes to a more sustainable and profitable recycling program.

If used in a consumer application for a resident of NYC, this model would have the following impacts: 
* Clearer and more specific guidelines on how to sort their waste items correctly, aiding in reducing improper recycling and the number of recyclable items which end up in landfills 
* Understanding proper waste sorting practices becomes more accessible and user-friendly.
* Clear instructions empower individuals to make environmentally conscious decisions.

## Next Steps 
The methods use to tune the model could be improved if the analysis were to be done again. We may explore Cross Validation with more parameters for KNN and Decision Tree models as was done with the SVM models. 

For example, the KNN models have many parameters that can be tuned including: `n_neighbors`, `weights`, `metric`, `algorithm`, `leaf_size`, `p`, and `n_jobs`, the weighting scheme, distance metric, and many more. These paramaters would definitely result in a more fine-tuned parameters. 

## Conclusion
Overall, properly sorted waste leads to better recycling rates, reducing the overall environmental footprint and contributing to a greener city.

By implementing efficient waste sorting technologies, NYC's Waste Department can streamline operations, improve recycling rates, and foster a more environmentally conscious community. 